### PR TITLE
Add getUnique and and associated functions

### PR DESCRIPTION
**Pull request by cfraz89**
Add getUnique and and associated functions

### CODE REVIEW

**Review by hasufell — CHANGES_REQUESTED** [review]
This needs tests.

File: src/Data/IxSet/Typed.hs [hunk 1]
```diff
@@ -847,36 +884,60 @@ innerJoinUsing (IxSet _ ixs1) (IxSet _ ixs2) _ = f (access ixs1) (access ixs2)
...
+  atPrimaryKey' get
     i
     ((\x ->
         case x of
           Nothing -> pure Nothing
           Just y -> Just <$> y) .
      fmap afb)
+{-# INLINE ixPrimaryKey' #-}
+
+-- | Assuming the given index is a /primary key/, constructs a traversal to
```
Comment (kccqzy) [inline src/Data/IxSet/Typed.hs]: This docstring is now misleading. Primary keys are supposed to be unique.

Comment (kccqzy) [reply src/Data/IxSet/Typed.hs]: In fact, why not just change the `atPrimaryKey` and `ixPrimaryKey` to use the new `getUnique` dingus and not add `ixUniqueKey`?

**Review by kccqzy — CHANGES_REQUESTED** [review]
Like the changes to `getOne` and `getUnique`, but I don't think `ixUniqueKey` needs to exist.

File: src/Data/IxSet/Typed.hs [hunk 2]
```diff
@@ -514,22 +519,47 @@ insert = change Set.insert Ix.insert
...
-updateIx i new ixset = insert new $
-                     maybe ixset (flip delete ixset) $
-                     getOne $ ixset @= i
+updateIx = updateIx' getOne
+
+-- | Will replace the item with the given index of type 'ix'.
+-- Only works if there is at most one item with that index in the 'IxSet'.
+-- Will generate an error if there is more than one item with given index.
+updateUnique :: (Indexable ixs a, IsIndexOf ix ixs)
+         => ix -> a -> IxSet ixs a -> IxSet ixs a
```
Comment (hasufell) [inline src/Data/IxSet/Typed.hs]: I feel like API wise this is a bit rough. The caller may want to decide whether he wants to error on non-unique index. Options are:
* wrap it in `Maybe (IxSet ixs s)` or `Either (IxSet ixs s) (IxSet ixs s)`
* return a tuple `(Bool, IxSet ixs a)`
* create a dedicated type with a record selector for people who don't care about whether the update worked or not
* use `Monadthrow`

File: src/Data/IxSet/Typed.hs [hunk 3]
```diff
@@ -847,36 +884,60 @@ innerJoinUsing (IxSet _ ixs1) (IxSet _ ixs2) _ = f (access ixs1) (access ixs2)
...
+ixUniqueKey = ixPrimaryKey' getUnique
+{-# INLINE ixUniqueKey #-}
+
+-- | Assuming the given index is a /primary key/, constructs a lens to access
+-- the value associated with the primary key. The getting operation uses 'getEQ'
+-- and 'getOne' and the setting operation uses 'updateIx' or 'deleteIx'.
+-- Therefore, this /will/ violate lens laws if the given index is not actually a
+-- primary key.
+atPrimaryKey :: (Indexable ixs a, IsIndexOf ix ixs, Functor f) => ix -> (Maybe a -> f (Maybe a)) -> (IxSet ixs a -> f (IxSet ixs a))
+atPrimaryKey = atPrimaryKey' getOne
```
Comment (hasufell) [inline src/Data/IxSet/Typed.hs]: probably want to inline this too

File: src/Data/IxSet/Typed.hs [hunk 4]
```diff
@@ -514,22 +519,47 @@ insert = change Set.insert Ix.insert
 delete :: Indexable ixs a => a -> IxSet ixs a -> IxSet ixs a
 delete = change Set.delete Ix.delete
 
+-- | Internal implementation for update* family
+updateIx' :: (Indexable ixs a, IsIndexOf ix ixs)
+         => (IxSet ixs a -> Maybe a) -> ix -> a -> IxSet ixs a -> IxSet ixs a
+updateIx' get i new ixset = insert new $
```
Comment (cevantes) [inline src/Data/IxSet/Typed.hs]: getter?

File: src/Data/IxSet/Typed.hs [hunk 5]
```diff
@@ -601,6 +631,13 @@ getOne ixset = case toList ixset of
 getOneOr :: a -> IxSet ixs a -> a
 getOneOr def = fromMaybe def . getOne
 
+-- | Like getOne, but error if multiple items exist
+getUnique :: IxSet ixs a -> Maybe a
+getUnique ixset = case toList ixset of
+                    [x] -> Just x
+                    [] -> Nothing
+                    _ -> error "getUnique: Multiple items in ixset"
```
Comment (cevantes) [inline src/Data/IxSet/Typed.hs]: how do you handle this error at the call site?

File: src/Data/IxSet/Typed.hs (hunk 1 above)
Comment (cfraz89) [reply src/Data/IxSet/Typed.hs]: I wouldn't mind doing that, but there's a bunch of code which currently hums away, possibly silently ignoring duplicates if they happen. If I switch over, they'll be throwing errors now. Kind of scary.

File: src/Data/IxSet/Typed.hs (hunk 2 above)
Comment (cfraz89) [reply src/Data/IxSet/Typed.hs]: Ok, its been updated to use MonadThrow

**Comment by cfraz89** [comment]
> This needs tests.

Done

**Review by hasufell — APPROVED** [review]

File: src/Data/IxSet/Typed.hs (hunk 1 above)
Comment (cevantes) [reply src/Data/IxSet/Typed.hs]: Another option is to pull this scary feature into local IxSet utils, and keep the library API simple.

**Review by cevantes — APPROVED** [review]

File: src/Data/IxSet/Typed.hs (hunk 1 above)
Comment (ip1981) [reply src/Data/IxSet/Typed.hs]: > Another option is to pull this scary feature into local IxSet utils, and keep the library API simple.

This.

**Review by kccqzy — APPROVED** [review]

**Comment by hasufell** [comment]
merge?

**Comment by kccqzy** [comment]
Not a big fan of `MonadThrow` but it's fine. I can we can merge.

**Pull request merged**